protocol/v2/ssv/validator/non_committee_validator.go

```diff
@@ -155,10 +160,12 @@ func (ncv *CommitteeObserver) ProcessMessage(msg *queue.SSVMessage) error {
 	return nil
 }
 
-func (ncv *CommitteeObserver) getRole(msg *queue.SSVMessage, root [32]byte) convert.RunnerRole {
+func (ncv *CommitteeObserver) getRole(msg *queue.SSVMessage, slot phase0.Slot, validatorIndex phase0.ValidatorIndex) convert.RunnerRole {
 	if msg.MsgID.GetRoleType() == spectypes.RoleCommittee {
-		_, found := ncv.Roots[root]
-		if !found {
+		beacon := ncv.networkConfig.Beacon
+		period := beacon.EstimatedSyncCommitteePeriodAtEpoch(beacon.EstimatedEpochAtSlot(slot))
+		d := ncv.dutyStore.SyncCommittee.Duty(period, validatorIndex)
+		if d == nil {
```

Comment:
I was considering a similar approach checking for the existence of a duty.

**Pros:**
 - It’s a straightforward method to determine if a duty exists for a specific role.
 - Simplicity: avoids the need to store roots and differentiate roles based on them.
 
**Cons:**
 - It may introduce reorg/indices change anomalies, though these are very rare edge cases.

If we go with this direction, we’ll need to deprecate the root logic, as it wasn't implemented according to the author's intention. We should also consider implementing the original logic as intended to ensure it works as expected.

Reply:
We need to adjust the logic in the `getRole` function to return roles based on the quorum, since the same quorum of operators can execute both attester and sync committee duties at the same slot.

```go
func (ncv *CommitteeObserver) getRoles(
	msgRole spectypes.RunnerRole,
	slot phase0.Slot,
	validatorIndex phase0.ValidatorIndex,
) []convert.RunnerRole {
	roles := make([]convert.RunnerRole, 0)

	if msgRole == spectypes.RoleCommittee {
		beacon := ncv.networkConfig.Beacon
		epoch := beacon.EstimatedEpochAtSlot(slot)
		period := beacon.EstimatedSyncCommitteePeriodAtEpoch(epoch)
		if d := ncv.dutyStore.SyncCommittee.Duty(period, validatorIndex); d != nil {
			roles = append(roles, convert.RoleSyncCommittee)
		}
		if d := ncv.dutyStore.Attester.ValidatorDuty(epoch, slot, validatorIndex); d != nil {
			roles = append(roles, convert.RoleAttester)
		}
	} else {
		roles = append(roles, convert.RunnerRole(msgRole))
	}
	return roles
}
```

This way, we iterate over the returned roles and try to save participants for each of them, ensuring that duties for both attester and sync committee are correctly handled when they occur at the same slot.

Reply:
@nkryuchkov I agree with @olegshmuelov here. actually on both cases.
- We should be able to determine the roles by the roots. I'm not sure why its not working atm but since attester and sync committee have different signing roots, this should be achievable.
- If for some reason we're taking the `dutyStore` approach (IMO a bit more complex since like @olegshmuelov said, reorgs might occur or maybe the duties might be deleted for some other reason), we need to make sure we cover all roles for the slot.

@nkryuchkov if there's any reason why the roots approach can't be achieved lets discuss..